File: LeetCode/000001.py
from typing import List


class Solution:
    def twoSum_n2(self, nums: List[int], target: int) -> List[int]:
        for i, v in enumerate(nums):
            for j, v2 in enumerate(nums[i+1:], i+1):
                if v + v2 == target:
                    return [i, j]
    
    def twoSum_n(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}
        for i, v in enumerate(nums):
            complement = target - v
            if complement in hashmap:
                return [hashmap[complement], i]
            hashmap[v] = i

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # return self.twoSum_n2(nums, target)
        return self.twoSum_n(nums, target)

File: LeetCode/000001.go
package twoSum

func twoSum_n2(nums []int, target int) []int {
	totalNums := len(nums)
	result := make([]int, 2)
	for i := 0; i < totalNums; i++ {
		for j := i + 1; j < totalNums; j++ {
			if nums[i]+nums[j] == target {
				result[0] = i
				result[1] = j
				return result
			}
		}
	}
	return result
}

func twoSum_n(nums []int, target int) []int {
	totalNums := len(nums)
	hashmap := map[int]int{}
	result := make([]int, 2)
	for i := 0; i < totalNums; i++ {
		complement := target - nums[i]
		if _, exists := hashmap[complement]; exists {
			result[0] = i
			result[1] = hashmap[complement]
			break
		} else {
			hashmap[nums[i]] = i
		}
	}
	return result
}

func twoSum(nums []int, target int) []int {
	return twoSum_n(nums, target)
}
